src/lib/server/result_serializer.cpp

```diff
@@ -0,0 +1,122 @@
+  auto attribute_strings = std::vector<std::optional<std::string>>(table->column_count());
+
+  const auto chunk_count = table->chunk_count();
+
+  // Iterate over each chunk in result table
+  for (ChunkID chunk_id{0}; chunk_id < chunk_count; chunk_id++) {
+    const auto& current_chunk = table->get_chunk(chunk_id);
+    const auto chunk_size = current_chunk->size();
+    const auto& segments = current_chunk->segments();
+    // Iterate over each row in chunk
+    for (ChunkOffset current_chunk_offset{0}; current_chunk_offset < chunk_size; ++current_chunk_offset) {
+      auto string_length_sum = 0;
+      // Iterate over each attribute in row
+      for (size_t current_segment = 0; current_segment < segments.size(); current_segment++) {
+        const auto attribute_value = (*segments[current_segment])[current_chunk_offset];
+        // The PostgreSQL protocol requires the conversion of values to strings
+        const auto string_value = lossy_variant_cast<pmr_string>(attribute_value);
+        if (string_value.has_value()) {
+          // Sum up string lengths for a row to save an extra loop during serialization
+          string_length_sum += string_value.value().size();
+        }
+        attribute_strings[current_segment] = std::move(string_value);
+      }
+      postgres_protocol_handler->send_values_as_strings(attribute_strings, string_length_sum);
+    }
+  }
+}
+
+std::string ResultSerializer::build_command_complete_message(const OperatorType root_operator_type,
+                                                             const uint64_t row_count) {
+  switch (root_operator_type) {
+    case OperatorType::Insert: {
+      // 0 is ignored OID and 1 inserted row
+      return "INSERT 0 1";
+    }
+    case OperatorType::Update: {
+      // We do not return how many rows are affected, because we don't track this
+      // information
+      return "UPDATE -1";
+    }
+    case OperatorType::Delete: {
+      // We do not return how many rows are affected, because we don't track this
+      // information
+      return "DELETE -1";
+    }
+    default:
+      // Assuming normal query
+      return "SELECT " + std::to_string(row_count);
+  }
+}
+
+template void ResultSerializer::send_table_description<Socket>(std::shared_ptr<const Table>,
+                                                               const std::shared_ptr<PostgresProtocolHandler<Socket>>&);
+
+template void ResultSerializer::send_table_description<boost::asio::posix::stream_descriptor>(
+    std::shared_ptr<const Table>,
+    const std::shared_ptr<PostgresProtocolHandler<boost::asio::posix::stream_descriptor>>&);
+
+template void ResultSerializer::send_query_response<Socket>(std::shared_ptr<const Table>,
+                                                            const std::shared_ptr<PostgresProtocolHandler<Socket>>&);
+
+template void ResultSerializer::send_query_response<boost::asio::posix::stream_descriptor>(
```

Comment:
"serialization is the process of translating data structures or object state into a format that can be stored or transmitted". The process of sending itself is not really the job of a serializer. Any idea for a dfferent name?

Reply:
ResultHandler? ResultConverter? I'm out of ideas.

Reply:
I gave it some more though too and couldn‘t come up with anything good either. Let‘s leave it.